### PR TITLE
clean garbage dirs in tmp after composer tests #4694

### PR DESCRIPTION
**Pull request by chorry**

### CODE REVIEW

File: tests/Composer/Test/CacheTest.php [hunk 1]
```diff
@@ -12,15 +12,20 @@
...
 {
     private $files, $root, $finder, $cache;
+    /**
+     * @var Filesystem
+     */
+    private $fs;
 
     public function setUp()
     {
+        $this->fs = new Filesystem;
```
Comment (stof) [inline tests/Composer/Test/CacheTest.php]: no need to instantiate it on setUp and to store it in a property. Instantiate it only when necessary

File: tests/Composer/Test/Downloader/GitDownloaderTest.php [hunk 2]
```diff
@@ -18,6 +18,25 @@
...
+    /** @var Filesystem */
+    private $fs;
+    /** @var string */
+    private $workingDir;
+
+    protected function setUp()
+    {
+        $this->fs = new Filesystem;
+        $this->workingDir = realpath(sys_get_temp_dir()).DIRECTORY_SEPARATOR.'cmptest-'.md5(uniqid('', true));
+        $this->fs->ensureDirectoryExists($this->workingDir);
```
Comment (stof) [inline tests/Composer/Test/Downloader/GitDownloaderTest.php]: IMO, this line should be removed, as several tests don't need it (and tests needing it are actually already calling it to ensure that a subfolder exists)

File: tests/Composer/Test/Downloader/HgDownloaderTest.php [hunk 3]
```diff
@@ -17,6 +17,25 @@
...
+    /** @var Filesystem */
+    private $fs;
+    /** @var string */
+    private $workingDir;
+
+    protected function setUp()
+    {
+        $this->fs = new Filesystem;
+        $this->workingDir = realpath(sys_get_temp_dir()).DIRECTORY_SEPARATOR.'cmptest-'.md5(uniqid('', true));
+        $this->fs->ensureDirectoryExists($this->workingDir);
```
Comment (stof) [inline tests/Composer/Test/Downloader/HgDownloaderTest.php]: same here

File: tests/Composer/Test/Downloader/PerforceDownloaderTest.php [hunk 4]
```diff
@@ -30,10 +31,13 @@ class PerforceDownloaderTest extends \PHPUnit_Framework_TestCase
     protected $repoConfig;
     protected $repository;
     protected $testPath;
+    protected $fs;
 
     protected function setUp()
     {
+        $this->fs = new Filesystem;
         $this->testPath        = sys_get_temp_dir() . '/composer-test';
+        $this->fs->ensureDirectoryExists($this->testPath);
```
Comment (stof) [inline tests/Composer/Test/Downloader/PerforceDownloaderTest.php]: useless, as most tests don't need to have this folder. There is no reason to create it for nothing

File: tests/Composer/Test/Downloader/XzDownloaderTest.php [hunk 5]
```diff
@@ -13,14 +13,32 @@
...
+     * @var string
+     */
+    private $testName;
+
     public function setUp()
     {
         if (defined('PHP_WINDOWS_VERSION_BUILD')) {
             $this->markTestSkipped('Skip test on Windows');
         }
+        $this->fs = new Filesystem;
```
Comment (stof) [inline tests/Composer/Test/Downloader/XzDownloaderTest.php]: this should be instantiated only on teardown

File: tests/Composer/Test/Downloader/ZipDownloaderTest.php [hunk 6]
```diff
@@ -13,14 +13,32 @@
...
+     * @var string
+     */
+    private $testName;
+
     public function setUp()
     {
         if (!class_exists('ZipArchive')) {
             $this->markTestSkipped('zip extension missing');
         }
+        $this->fs = new Filesystem;
```
Comment (stof) [inline tests/Composer/Test/Downloader/ZipDownloaderTest.php]: same here

File: tests/Composer/Test/Downloader/ZipDownloaderTest.php [hunk 7]
```diff
@@ -13,14 +13,32 @@
...
 {
+    /**
+     * @var Filesystem
+     */
+    private $fs;
+
+    /**
+     * @var string
+     */
+    private $testName;
```
Comment (stof) [inline tests/Composer/Test/Downloader/ZipDownloaderTest.php]: ``testDir`` would be a better name than ``testName``

**Comment by alcohol** [comment]
It seems to randomly fail the `InstallerTest` cause of tmp directory issues. I restarted the build but then it just failed on another PHP version. Going to see if anyone from travis can give feedback on that :-\

**Comment by chorry** [comment]
I see travis is now marked as passed, but appveyour is failed due to failed provisioning of VM. 
Can somebody restart it?

**Comment by Seldaek** [comment]
Seems fine on travis and on my machine so let's hope appveyor will go better in the future, thanks for the fix!

**Pull request merged**